File: src/main/java/linq/ToCollection.java
package linq;

import linq.exception.ExceptionArgument;
import linq.exception.ThrowHelper;

import java.util.ArrayList;
import java.util.List;

public class ToCollection {
    static <TSource> List<TSource> toList(Enumerable<TSource> source) {
        if (source == null) {
            ThrowHelper.throwArgumentNullException(ExceptionArgument.source);
        }

        if(source instanceof AbstractIterator<TSource> iterator) {
            return iterator.toList();
        }

        var list = new ArrayList<TSource>();
        var iterator = source.iterator();
        iterator.forEachRemaining(list::add);
        return list;
    }
}

File: src/main/kotlin/linq/ToCollection.kt
package linq

import linq.exception.ExceptionArgument
import linq.exception.throwArgumentNullException

fun <TSource> toList(source: Enumerable<TSource>?): List<TSource> {
    if (source == null) {
        throwArgumentNullException(ExceptionArgument.Source)
    }

    if (source is AbstractIterator<TSource>) {
        return source.getList()
    }

    val list = arrayListOf<TSource>()
    for (item in source) {
        list.add(item)
    }
    return source.toMutableList()
}
